// app/back/preload/safe-ipc.js
/* safe-ipc.js
    
    Whitelisted ipcRenderer object.
*/

// Dependencies.
const { contextBridge, ipcRenderer } = require("electron");

/* Validates a channel against a given whitelist.
    
    channel - IPC channel string to validate.
    whitelist - Object containing whitelisted channel strings as keys. All
        values must be truthy.
    
    Returns true if channel is whitelisted, or throws an error if not
    whitelisted.
*/
function validateChannel(channel, whitelist) {
    if (!whitelist[channel]) {
        throw new Error(`Channel '${channel}' is not valid.`);
    }
    
    return true;
}

/* Calls a given method on the IPC renderer.
    
    method - Method to call on the ipcRenderer object.
    channel - Whitelisted IPC channel string.
    listener - Optional listener associated with the channel.
*/
function ipcCallMethod(method, whitelist, channel, listener) {
    validateChannel(channel, whitelist);
    ipcRenderer[method](channel, listener);
}

/* Creates a context bridge to invoke/handle whitelisted ipcRenderer channels.
    
    channels - String or array of string IPC channels to whitelist.
    name - Optional object name to use on the BrowserWindow window object.
        Defaults to 'ipc'.
*/
module.exports = (channels, name) => {
    channels = typeof channels === "string" ? [ channels ] : channels;
    name = (typeof name === "string" ? name : "ipc").trim();
    
    if (!(channels instanceof Array)) {
        throw new Error("Specified channels are invalid or missing.");
    }
    
    if (!name) {
        throw new Error("Specified object name is invalid (must be a non-" +
            "empty string with at least 1 non-whitespace character).");
    }
    
    // Convert the channels array into a whitelist object
    const whitelist = channels.reduce((obj, curr) => {
        curr = (typeof curr === "string" ? curr : "").trim();
        curr && (obj[curr] = true);
        
        return obj;
    }, {});
    
    Object.freeze(whitelist);
    
    if (!Object.keys(whitelist).length) {
        throw new Error("No valid channel strings provided (must be non-" +
            "empty strings with at least 1 non-whitespace character).");
    }
    
    // Create the safe helper object
    contextBridge.exposeInMainWorld(
        name,
        {
            /* Sends an asynchronous message to a whitelisted IPC channel. To
                recieve a response from the IPC main handler, use this.invoke()
                instead.
                
                channel - IPC channel string to which to send the message.
                ...args - Arguments to send to the IPC channel.
            */
            send: (channel, ...args) => {
                validateChannel(channel, whitelist);
                ipcRenderer.send(channel, ...args);
            },
            
            /* Invokes a whitelisted IPC channel.
                
                channel - IPC channel string to invoke.
                ...args - Arguments to pass to the IPC channel.
            */
            invoke: async (channel, ...args) => {
                validateChannel(channel, whitelist);
                return await ipcRenderer.invoke(channel, ...args);
            },
            
            /* Attaches a listener to a whitelisted IPC channel.
                
                channel - IPC channel string to attach the listener.
                listener - Listener function to attach to the channel.
            */
            on: (channel, listener) => {
                ipcCallMethod("on", whitelist, channel, listener);
            },
            
            /* Attaches a one-time listener to a whitelisted IPC channel.
                
                channel - See this.on().
                listener = See this.on().
            */
            once: (channel, listener) => {
                ipcCallMethod("once", whitelist, channel, listener);
            },
            
            /* Removes a listener from a whitelisted IPC channel.
                
                channel - See this.on().
                listener = See this.on().
            */
            removeListener: function(channel, listener) {
                ipcCallMethod("removeListener", whitelist, channel, listener);
            },
            
            /* Removes all listeners from a whitelisted IPC channel.
                
                channel - See this.on().
                listener = See this.on().
            */
            removeAllListeners: function(channel) {
                ipcCallMethod("removeAllListeners", whitelist, channel);
            }
        }
    );
};
