// app/front/output/components/ingredients-card.js
/* ingredients-card.js
    
    Recipe ingredients card.
*/

// Dependencies.
const Card = require("app/front/output/components/card.js");

/* Recipe ingredients card component.
*/
class IngredientsCard extends Card {
    //
    // STATIC PROPERTIES
    //
    
    //
    // PROPERTIES
    //
    
    // Reference to the card's top-level element.
    _elem = null;
    
    /* Unmount handler.
    */
    componentWillUnmount() {
        super.componentWillUnmount();
        
        this._elem = null;
    }
    
    //
    // STATIC METHODS
    //
    
    //
    // METHODS
    //
    
    /* Getter for _elem.
        
        Returns the top-level element for the component.
    */
    get elem() {
        return this._elem;
    }
    
    /* Gets an author DOM element for the card, if an author is defined.
    */
    getAuthorDom(author) {
        author = (typeof author === "string" ? author : "").trim();
        
        if (!author) {
            return null;
        }
        
        let possessive = String.fromCharCode(8217) +
            (author.match(/s$/i) ? "" : "s");
        
        return (
            <h3 className="col-12">{`${author}${possessive} Recipe`}</h3>
        );
    }
    
    /* Gets DOM elements for the ingredients list.
        
        ingredients - Ingredient string or array of strings. Falsy and
            non-string values will be skipped.
        
        Returns DOM structure for the ingredients list.
    */
    getIngredientsDom(ingredients) {
        ingredients = ingredients instanceof Array ?
            ingredients : [ ingredients ];
        
        ingredients = ingredients.filter((v) => {
            return (v && typeof v === "string");
        });
        
        let list = [],
            cutoff = ingredients.length > 5 ?
                Math.ceil(ingredients.length / 2) : ingredients.length,
            timestamp = Date.now();
        
        for (let i=0, l=ingredients.length; i<l; ++i) {
            let elem = <li
                key={`timestamp-${i}`}
                style={i === cutoff ? { breakBefore:"column" } : {}}
            >{ingredients[i]}</li>;
            
            list.push(elem);
        }
        
        return list;
    }
    
    /* Renders the component.
    */
    render() {
        let author = this.getAuthorDom(this.props.author),
            ingredients = this.getIngredientsDom(this.props.ingredients);
        
        return (
            <article
                className="recipe-card ingredients grid"
                ref={(r) => { this._elem = r; }}
            >
                
                <header className="row">
                    <h1 className="col-6">{this.props.name}</h1>
                    <h2 className="col-6">Ingredients</h2>
                    {author}
                </header>
                
                <section
                    className="columns col-12"
                    ref={(r) => { this._resizeElem = r; }}
                >
                    
                    <ul className="allow-break">
                        {ingredients}
                    </ul>
                    
                </section>
                
            </article>
        );
    }
}

module.exports = IngredientsCard;
