// app/front/browser/components/browser-form.js
/* browser-form.js
    
    Recipe browser form component.
*/

// Dependencies.
const { Component } = require("react");

/* Recipe browser form component.
*/
class BrowserForm extends Component {
    //
    // STATIC PROPERTIES
    //
    
    //
    // PROPERTIES
    //
    
    /* Constructor for BrowserForm.
    */
    constructor(props) {
        super(props);
    }
    
    /* Mount handler.
    */
    componentDidMount() {
        this.setState({
            data: {}
        });
    }
    
    /* Destructor.
    */
    componentWillUnmount() {
        this.state.data = null;
    }
    
    /* Update handler.
    */
    componentDidUpdate() {
    }
    
    //
    // STATIC METHODS
    //
    
    //
    // METHODS
    //
    
    /* Form button click handler.
        
        e - Click event.
    */
    onClick = (e) => {
        e.preventDefault();
        e.stopPropagation();
        
        ipc.send("open-file");
        
        e.target.blur();
    };
    
    /* Renders the browser form.
    */
    render() {
        return (
            <form id="browser-form" className="row">
                <div className="col-12">
                    <input
                        id="receipe-file-button"
                        type="button"
                        accept=".recipe-pack"
                        value="Open Recipes&hellip;"
                        onClick={this.onClick}
                    />
                </div>
            </form>
        );
    }
}

module.exports = BrowserForm;
