// app/front/browser/components/browser-form.js
/* browser-form.js
    
    Recipe browser form component.
*/

// Dependencies.
const { Component } = require("react");

/* Recipe browser form component.
*/
class BrowserForm extends Component {
    //
    // STATIC PROPERTIES
    //
    
    //
    // PROPERTIES
    //
    
    // File button element.
    _fileBtn = null;
    
    // File input element.
    _fileInput = null;
    
    /* Constructor for BrowserForm.
    */
    constructor(props) {
        super(props);
    }
    
    /* Mount handler.
    */
    componentDidMount() {
        this.setState({
            data: {}
        });
    }
    
    /* Destructor.
    */
    componentWillUnmount() {
        this.state.data = null;
    }
    
    /* Update handler.
    */
    componentDidUpdate() {
    }
    
    //
    // STATIC METHODS
    //
    
    //
    // METHODS
    //
    
    /* Form button click handler.
        
        e - Click event.
    */
    onClick = (e) => {
        e.preventDefault();
        
        if (null === this._fileInput) {
            return;
        }
        
        this._fileInput.click();
    };
    
    /* File load change handler.
        
        e - Change event.
    */
    onFileLoad = async (e) => {
        e.preventDefault();
        
        let input = e.target,
            file = input.files.length ? input.files[0] : null,
            data;
        
        if (!file) {
            return;
        }
        
        // Load and parse
        try {
            data = await file.text();
        }
        catch (e) {
            throw new Error("Recipe file could not be loaded.");
            data = { recipes: [] }
        }
        
        try {
            data = JSON.parse(data);
        }
        catch (e) {
            throw new Error("Recipe file format is invalid.");
        }
        
        this._fileBtn.blur();
        
        this.props.loadRecipes(data);
    };
    
    /* Renders the browser form.
    */
    render() {
        return (
            <form id="browser-form" className="grid">
                <div className="col-12">
                    <input
                        id="receipe-file-button"
                        type="button"
                        accept=".recipe-pack"
                        value="Open Recipes&hellip;"
                        ref={(r) => { this._fileBtn = r; }}
                        onClick={this.onClick}
                    />
                    <input
                        id="recipe-file"
                        type="file"
                        className="hidden-file-input"
                        ref={(r) => { this._fileInput = r; }}
                        onChange={this.onFileLoad}
                    />
                </div>
            </form>
        );
    }
}

module.exports = BrowserForm;
