// app/front/browser/components/recipe-list-item.js
/* recipe-list-item.js
    
    Recipe list item component.
*/

// Dependencies.
const { Component } = require("react"),
    Recipe = require("app/front/recipe.js");

/* Recipe component.
*/
class RecipeListItem extends Component {
    //
    // STATIC PROPERTIES
    //
    
    //
    // PROPERTIES
    //
    
    /* Constructor for RecipeListItem.
    */
    constructor(props) {
        super(props);
        
        this.state = {
            recipe: new Recipe(props)
        };
    }
    
    /* Mount handler.
    */
    componentDidMount() {
        this.recipe = this.props;
    }
    
    /* Destructor for Recipe.
    */
    componentWillUnmount() {
        this.recipe.destroy();
    }
    
    //
    // STATIC METHODS
    //
    
    //
    // METHODS
    //
    
    /* Getter for state.recipe.
        
        Returns the recipe data object for the list item.
    */
    get recipe() {
        return this.state.recipe;
    }
    
    /* Setter for state.recipe.
        
        data - Raw recipe data.
    */
    set recipe(data) {
        this.setState({
            recipe: new Recipe(data)
        })
    }
    
    /* Renders the recipe.
    */
    render() {
        let recipe = this.state.recipe;
        
        return (
            <li key={recipe.name}>
                <div className="col-8">
                    <span className="title">{recipe.title}</span>
                </div>
                <div className="controls col-4">
                    <div className="delete col-1">X</div>
                    <div className="move col-1">=</div>
                </div>
            </li>
        );
    }
}

module.exports = RecipeListItem;
