// app/front/browser/components/recipe-list-item.js
/* recipe-list-item.js
    
    Recipe list item component.
*/

// Dependencies.
const { Component } = require("react"),
    Recipe = require("app/front/recipe.js");

/* Recipe component.
*/
class RecipeListItem extends Component {
    //
    // STATIC PROPERTIES
    //
    
    //
    // PROPERTIES
    //
    
    /* Constructor for RecipeListItem.
    */
    constructor(props) {
        super(props);
        
        this.state = {
            recipe: new Recipe(props)
        };
    }
    
    /* Mount handler.
    */
    componentDidMount() {
        this.recipe = this.props;
    }
    
    /* Destructor.
    */
    componentWillUnmount() {
        this.recipe.destroy();
    }
    
    //
    // STATIC METHODS
    //
    
    //
    // METHODS
    //
    
    /* Getter for state.recipe.
        
        Returns the recipe data object for the list item.
    */
    get recipe() {
        return this.state.recipe;
    }
    
    /* Setter for state.recipe.
        
        data - Raw recipe data.
    */
    set recipe(data) {
        this.setState({
            recipe: new Recipe(data)
        })
    }
    
    /* List item click handler.
        
        e - Click event.
    */
    onClick = (e) => {
        e.preventDefault();
        
        let elem = e.currentTarget,
            selected = elem.parentElement.querySelectorAll(".selected");
        
        for (let i=0, l=selected.length; i<l; ++i) {
            selected[i].classList.remove("selected");
        }
        
        elem.classList.add("selected");
        
        ipc.send("select-recipe", this.state.recipe.id);
    };
    
    /* Renders the recipe.
    */
    render() {
        let recipe = this.state.recipe;
        
        return (
            <li
                key={recipe.id}
                className="row"
                onClick={this.onClick}
            >
                <div className="name col-9">
                    <span className="name">{recipe.name}</span>
                </div>
                <div className="controls col-1">
                    <div className="print"><i className="fas fa-print"></i></div>
                </div>
                <div className="controls col-1">
                    <div className="delete"><i className="fas fa-times"></i></div>
                </div>
                <div className="controls col-1">
                    <div className="move"><i className="fas fa-sort"></i></div>
                </div>
            </li>
        );
    }
}

module.exports = RecipeListItem;
