// app/front/browser/recipe-list.js
/* recipe-list.js
    
    Recipe list object.
*/

// Dependencies.
const LinkedList = require("app/all/linked-list.js"),
    Recipe = require("app/front/recipe.js"),
    dataMap = require("app/all/data-map.js");

/* Recipe list object.
*/
class RecipeList {
    //
    // STATIC PROPERTIES
    //
    
    //
    // PROPERTIES
    //
    
    // Internal recipe array.
    _list = new LinkedList();
    
    // Top-level recipe list element.
    _elem = null;
    
    /* Constructor for RecipeList.
        
        recipes - Array of recipe data objects.
        elem - Placeholder element for the recipe list DOM.
    */
    constructor(recipes, elem) {
        if (recipes) {
            recipes = recipes instanceof Array ? recipes : [ recipes ];
        }
        else {
            recipes = [];
        }
        
        for (let i=0, l=recipes.length; i<l; ++i) {
            this.addRecipe(recipes[i]);
        }
        
        if (elem) {
            this.elem = elem;
            this.render();
        }
    }
    
    /* Destructor for RecipeList.
    */
    destroy() {
        this._list.destroy();
        this._list = null;
    }
    
    //
    // STATIC METHODS
    //
    
    //
    // METHODS
    //
    
    /* Getter for _list.
        
        Returns the internal array of recipes.
    */
    get list() {
        return this._list;
    }
    
    /* Setter for _list (don't overwrite the list array directly, this will
        break things).
        
        unused - Unused.
    */
    set list(unused) {}
    
    /* Getter for _elem.
        
        Returns the top-level recipe list DOM element.
    */
    get elem() {
        return this._elem;
    }
    
    /* Setter for _elem.
        
        elem - New top-level recipe list DOM element.
    */
    set elem(elem) {
        if (!(elem instanceof Node)) {
            throw new Error("Recipe list element must be a valid DOM Node.");
            return;
        }
        
        if (null === elem.ownerDocument) {
            throw new Error("Recipe list element cannot be a Document " +
                "element.");
        }
        
        this._elem = elem;
    }
    
    /* Getter for _document.
        
        Returns the document element for the recipe list.
    */
    get document() {
        return this._elem.ownerDocument;
    }
    
    /* Adds a recipe to the list.
        
        recipe - Raw recipe data to add to the list.
        
        after - Optional existing recipe object in the list after which to add
            the new recipe. If omitted, the new recipe will be appended to the
            end of the list.
        
        Returns the list node containing the newly-added recipe.
    */
    addRecipe(recipe, after) {
        if (!(recipe instanceof Recipe)) {
            recipe = new Recipe(recipe);
        }
        
        let node;
        
        if (after instanceof Recipe) {
            node = this._list.insert(recipe, after);
        }
        else {
            node = this._list.push(recipe);
        }
        
        return node;
    }
    
    /* Removes a recipe from the list.
        
        node - List node containing the recipe to remove. Node must belong to
            this recipe list.
    */
    removeRecipe(node) {
        this._list.remove(node);
    }
    
    /* Renders the recipe list in place of the current element.
    */
    render() {
        if (!this.elem) {
            return;
        }
        
        let d = this.document,
            elem = d.createElement("ol");
        
        elem.classList.add("recipe-list", "grid");
        
        this._list.each((recipe, node) => {
            dataMap.set(recipe.renderListItem(elem), "node", node);
        });
        
        this._elem.replaceWith(elem);
        this.elem = elem;
    }
}

module.exports = RecipeList;
